### DISCUSSION TITLE
Need help understanding total number of Dags oscillating on UI

### DISCUSSION THREAD

**Discussion by rtrindvg** [apache/airflow#44495] [Q&A] [answered]
I am in the middle of a migration from an Airflow in a Virtual Machine to a kubernetes cluster, for now in a staging environment. After a lot of configuration adjustments on the Helm values.yaml, the cluster seems to be stable and working fine. 

But for some reason, the UI sometimes shows less DAGs than what are available. For example, we have a total of 93 DAGs. After the initial load, which takes a couple of minutes, it becomes stable for some time. Than it reduces to a smaller number (like 64) and after a couple of minutes, it starts to go back up again, eventually returning to 93 again. We confirmed this is not any kind of browser cache. There were no restarts of any pods in the meantime, no changes to the cluster and no DAGs were changes as well. 

We are using git-sync in a non-persistent storage, like it's recommended in the docs. We activated the debug logs on it and it seems to be working fine, just downloading changes when the DAGs branch has changes, and they seem to be propagating quickly to all relevant pods.

The scheduler logs were not clear on any kind of errors which could justify the drop in total DAGs, except the following line:
```
[2024-11-30T02:10:03.298+0000] {scheduler_job_runner.py:1782} INFO - Found (8) stales dags not parsed after 2024-11-30 02:00:03.296796+00:00.
```
I am researching if this is relevant to the issue at hand, but unsuccessfully so far.

Another fix we tried was activating the non-default DAG processor, but the behavior is the same. I tried activating the processor verbose mode using the env parameter, unsuccessfully. The logs are mostly blank, so I have no clue if the DAG processor is the culprit.

We also replaced the CeleryExecutor to the KubernetesExecutor, because it is more suited to our purposes. We did not think it had any relation to the issue and, as expected, the behavior persists.

Since I am from the cloud-infra team, and have no previous experience in Airflow, can someone help me understand what could be the issue and possible next steps in diagnosing our environment?

We are using airflow 2.9.3 (since it's the most recent in the latest Helm available), python 3.12, in a custom Dockerfile. We are not extending the image, we are really customizing it, since we need to perform a couple of compilations and it was more optimal to do this prior to the airflow pip installs, to make the rebuild faster and the final image smaller. I did not know if it was safe to point the image to the latest Airflow available (since I assume an updated Helm would be published if this was the case), so we kept using it this one.

Embedding the DAGs onto the image is not an option, since they are changed constantly and the time to rebuild and the process of redeploying the cluster several times a day is not ideal for us. If updating the cluster to 2.10.3 is safe and has any known issues regarding this behavior, please point me in the right way.

Thanks for any tips!

**Comment by potiuk** [comment, marked as answer]
> [2024-11-30T02:10:03.298+0000] {scheduler_job_runner.py:1782} INFO - Found (8) stales dags not parsed after 2024-11-30 02:00:03.296796+00:00.

That line is important. It tells that dag processor, did not parse some of your DAGs within https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#dag-stale-not-seen-duration  time (10 minutes) - the timestamps in your log file confirm that - simply 10 minutes passed since the parsing of the DAG files produced the 8 DAG ids that were deactivated.

## How you can prepare for investigation

Now. The question is why and we need you to investigate several things to find out.

1) First of all - find out which dags disappeared (got deactivated) - that will make further investigations easier.. Those are DAGs in the DAG table of airflow that will have "active = False". This is what is going to change in your case - so all the dags that are "non-active"  - will disappear from the UI.

2) The important thing here is that there is no 1-1 relationship between DAG files and DAG ids. Sometimes (Dynamic DAG generation) - parsing one DAG file might produce more than one DAG id. This is Dynamic DAG Generation. So you need to find out  which of the FILES in DAG folder should have produced the DAG ids that got deactivated.

3) Once you know that, there are a number of things that could go wrong.

## Things that could be wrong

**a) Your Dynamic DAG generation in your DAGs is buggy / unstable and produces different set of DAG ids every time it is parsed.**

Generally speaking, when you do Dynamic DAG generation, every time a DAG file is parsed, it shoudl produce consistently the same dag ids. Again - it is entirely up to you how the DAGs are written. There are different techniques you can use for dynamic DAG generation https://airflow.apache.org/docs/apache-airflow/stable/howto/dynamic-dag-generation.html#dynamic-dags-with-external-configuration-from-a-structured-data-file - and it might be that the way how you do it is simply unstable. For example pulls information about the various DAGs generated as Json file and that file content changes in unpredictable / non stable ways. Or there is a bug in the DAG generation code that causes an exception sometimes and not all DAGs are generated. Generally - when you do `python your_dag_file.py` - parsing it should consitently create the same number of DAG "objects" created in python globals with the same ids.

This is most probable cause.

**b) some of your files are not parsed for some reasons. Generally airflow DAG file processor is parsing continuously all files and should go through all the files in a loop, but there are certain parameters that control it:**

* https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#file-parsing-sort-mode - determines what is the sorting criteria used. 
* https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#file-parsing-sort-mode - this is how often we check if certain dags were last parsed to see if the https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#dag-stale-not-seen-duration has been exceeded,
* This is how many parallel parsing processes are run by each scheduler (or dag file process) if you have standalone dag file processor) https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#parsing-processes

There are few other parameters, but those are the important ones. 

**c)  Your DAGs do notfolow best practices.**

So one of other options is that simply parsing of some of your DAG files takes a long time - long enough that it takes more than 1- minutes for the dag file processor to go in a look to parse all your files. If you follow best practices https://airflow.apache.org/docs/apache-airflow/stable/best-practices.html#reducing-dag-complexity - and make sure that you do not block or spend a lot of time in parsing your top-level code of DAG files https://airflow.apache.org/docs/apache-airflow/stable/best-practices.html#top-level-python-code  - parsing of even 100s of files should take seconds at most. But there are cases - especially when your top-level code is reaching out to external sources and possibly "hangs" or takes a lot of time to complete parsing, your parsing can take arbitrary long - minutes or hours. 

If you have some of your DAG files doing it, they might simply hang for a long time - if you have 2 parallel dag parsing processes running - it's enough that you have two files where parsing takes around 10 minutes and that will cause some of the remaining files to not be parsed "on time".  Similarly if you have many DAG files that are parsed in few minutes each - that might delay the queue enough, that some of your DAGs will not parse within default 10 minutes

You need to make sure to optimize your parsing time and follow best practices - ideally, all your DAGs should be parsed in seconds rather than minutes. Of course you can also play with the parameters above, and make timeouts bigger, but that is rather masking the "long parsing being problematic" rather than solve the problem and will results in for example far longer times on reflecting dag file changes into parsed DAGs in DB. 

You can control DAG parsing timeout https://airflow.apache.org/docs/apache-airflow/stable/faq.html#how-to-control-dag-file-parsing-timeout-for-different-dag-files and the next section https://airflow.apache.org/docs/apache-airflow/stable/faq.html#when-there-are-a-lot-1000-of-dag-files-how-to-speed-up-parsing-of-new-files explains some of the ways you can attempt speed up parsing.

This is the second most probable cause for what you see.

**d) your syncing process might have some on/off states where the files are appearing / disappearing after check-out.** 

Git-sync works in the way that it checks out the latest commit and then swaps out with previous check-out via symbolic link. Maybe there is 
somethign in that process  that causes it? Maybe for example ther are some permission problems that prevent the files to be parsed etc. etc. 

Not very likely though

**e) it might be that some of the combination of the parameters above and your synchronization settings above causes bad sorting of parsed files**

It might simply cause that the processors are not parsing your files - there might be various reasons - for example if your sorting order is `modified time` and for some reason your syncing process causes mtime to be modified continuously, it might well be that dag processor will only ever attempt to parse "last modified" files and your "non-modified files" will be always put at the end of the queue - that might cause similar issues - you must look for an indication (in dag file processor logs) what files are being parsed. This might also happen if the filesystem of yours  badly handles mtime coming from git - or when your git is configured to not preserve modified time that is coming from git. 

This is also quite unlikely reason, but if you use non POSIX compliant shared filesystems, i can imagine it can happen.

**f) Finally - time on your various machines might not be synchronized.**

If one of the machines (DB, scheduler, processor, git)  has a significant drift of time, it might be that causes various time calculation wrong. 

Not very likely, most of the computing resources out there have ntp and similar way of syncing time, but we've seen that happening.

Good luck with your investigation and please come back here with the results.

**Reply by rtrindvg** [reply]
This description is a godsend, @potiuk, thank you so much for the time to put all this relevant information in a single topic! Most of what you say here I read in several different places of the documentation, but have it all here in a single explanation is just mega helpful and I appreciate it a lot.

I will pass this information to the team at Monday and will come back with the results. I am certain with this information we will be able to solve the issue and I will report back my findings here, to help more people which may fall in the same scenario.

**Comment by potiuk** [comment]
As explained in #39332

@vghar-bh
@raphaelauv
@AlexHAM
@rtrindvg

Please follow the above guide and explain results of your investigation that I outlined - that might help us with maybe getting better messaging and explanation on what would go wrong with a number of deployments - or maybe we will eventually find some problem in Airflow that might cause it. 

I'd really appreciate a dilligent reading and understanding of how things work and how it maps to your deployment, because this is the only way we can get to the bottom of the issues you have with disappearing DAGs - that might help us to build more resilience, messaging and possibly even find some bugs in Airflow that might help other users having similar issues.

Your help here and dilligence is greatly appreciated.

**Reply by rtrindvg** [reply]
In my specific case, I am not the main user of the airflow tool, just an infra guy, so in my case, I do not put blame in the docs at all. I am parsing through the documents for the last couple of weeks, but always focused of changing a specific behavior of the platform.

From what I could feel, airflow docs are very good and I had no difficulty customizing the image, deploying the official Helm file and changing configurations to reach the desired outcome. My main difficulty was the time it took to find the line which could explain the problem. I just found it yesterday, and a quick google did not bring any relevant results right away, but I am sure with more time we could reach the solution, eventually.

**Comment by potiuk** [comment]
BTW. I am plannig to turn that above explanation of mine into "What to do if you see disappearing DAGs" FAQ on our website, so any feedback to that description is also most welcome.

**Reply by rtrindvg** [reply]
I think it makes a lot sense. It would help cases like mine right away and although I won't have a chance to try the suggestions until monday, I am sure it will already help a lot of people in the way it is structured right now.

Thanks again for all the help!

**Comment by rtrindvg** [comment]
Found the issue in our case. It was option "c". After some discussion with the data team, which is the one which creates the DAGs, it was clear no best practices were being used. Besides large imports (which did not need to happen during the parse stage), there were some costly operations happening, like connecting to a vault to get secure parameters and also some database connections. This multiplied by the number of affected DAGs clearly explained the observed behavior.

Thanks again for the detailed post which made my life a lot easier for them to understand it was not an infrastructure issue.

**Reply by potiuk** [reply]
> Thanks again for the detailed post which made my life a lot easier for them to understand it was not an infrastructure issue.

Cool. Thanks! so now it's the time to turn it into a FAQ... Or maybe you could help with that and become contributor @rtrindvg ? It's very easy - go to the page with FAQS https://airflow.apache.org/docs/apache-airflow/stable/faq.html - click "suggest a change on this page" and it will open PR where you could turn my answer into a FAQ entry and lead it to completion (likely some typos/ .rst formatting etc. ) 

Maybe that's a good form of "thanks" and also paving the path for future people like you :)

WDYT ?

**Reply by rtrindvg** [reply]
Mission accepted!

Probably won't be done quickly because of some personal challenges and holidays, but as soon as I am able I will open this PR. Since I am not a native speaker, I will kindly ask for someone to later review the text and make sure everything is as it should be.

**Reply by potiuk** [reply]
I am not a native speaker myself and a lot of what I write is corrected by others - you can count on that. We have a few grammar-conscious people here. Also docs are verified via spell-checker :).

**Reply by rtrindvg** [reply]
It took a while, but it is finally done. PR opened: https://github.com/apache/airflow/pull/49987

**Comment by jradhima** [comment]
Hello all, we recently encountered such an issue in our setup. Briefly:

- deployment on k8s using the official helm chart, 4+ schedulers, 4K+ dags
- parsing takes approx 25 minutes with `min_file_process_interval` set to 50 minutes
- moved to standalone dag processor and dags were disappearing and reappearing from the UI

After reading this post, we looked into the interaction between scheduler and dag processor and we found out that the issue was:
- parser parses every 50 minutes
- dag_stale_not_seen_duration is set to default 10 minutes
- scheduler removes all dags 10 minutes after the processor parses them

Our solution for the moment is to set `dag_stale_not_seen_duration` equal to `min_file_process_interval` + 1h. 

However, we do have this question: the same setup does not cause any issues when the schedulers are parsing the dags - is this a "bug"/unforeseen behaviour of the  setup, or is this considered standard behaviour that all users should be familiar with and this works as expected?

To generalise the above, after looking into the source code and the logs more closely we realised that both the scheduler and the dag processor are responsible for removing "stale" dags in different ways. My opinion is that this can cause confusion because when moving the parsing of the dags into a standalone component, it makes sense to move everything regarding dags, serialised dags & their management into that component.

---

Tangential to the above, allow me to pose another question: (to what extent) does running multiple dag processing managers end up in duplicate work and under what circumstances? I've seen many logs and indications that the processing managers work independently, however it doesn't make sense that they do not communicate with the database at all, so there has to be some interaction through the "last_parsed" attribute of the `serialized_dag` table.

 My current understanding is:
- each processing manager will make a file queue using the last modified time (from the db) and checking against min_file_process_interval
- until the file queue is cleared, the manager will not communicate with the database again, the manager does not check the db on a per-dag or per-file basis
- it will bulk-update the database with the parsing results in the end
- this means that if a file has been picked up by a processor manager, it will remain "open" for others until the manager bulk syncs

I am sure that this is not exactly how the process works but it should contain elements of truth in it. I know that processors do *some* duplicate work, what I am unsure is if they always do duplicate work. I think this area is lacking in the docs and would be open to help clarify it, however I would first need to understand it completely myself!

**Reply by pykenny** [reply]
From [`dag_stale_not_seen_duration`'s documentation](https://airflow.apache.org/docs/apache-airflow/2.10.5/configurations-ref.html#dag-stale-not-seen-duration):

> Only applicable if [scheduler] standalone_dag_processor is true.

And that's why you didn't see this phenomenon when you use schedule subprocess to parse DAGs (disable `standalone_dag_processor`), as long as DAG definition files are still in the file system and get refreshed by the subprocess parser.

**Reply by pykenny** [reply]
When scheduler and DAG processing runs independently  (enable `standaline_dag_processor`), scheduler has no access to DAG definition folders ~~, and need some protocol between DAG processor to know when a DAG definition gets removed from the file system and should be deactivated~~. Standalone DAG processor tries its best to ~~send the information (discovering new ones/updating existing ones/removal)~~ update DAG Bag for the scheduler but still has some limitation, hence `dag_stale_not_seen_duration` is a bottom line to make final decision when DAG scheduler stops receiving any information about some managed DAG from DAG processor.

`min_file_process_interval`, on the other hand, is the last line of defense for DAG parser to ensure a DAG definition's freshness by force parsing definition file constantly, in case it fails to detect change in the file system correctly.

**Reply by jradhima** [reply]
Thanks for the reply. I understand the configuration settings, I just don't understand the design decision taken.

The point of the standalone dag processor is that the scheduler only schedules and doesn't need to have access to the dag files. Why is airflow bypassing this by faking access to this information? Shouldn't the dags be a responsibility of the processor and the scheduler just schedules?

I think it's more intuitive that when a dag file is removed or the source code of a dag is modified and the dag becomes stale, the dag processor removes it instead of the scheduler. This way, if you have any issues, you only need to check one place in the code/logs/configuration to figure out the behavior.

**Reply by pykenny** [reply]
Sorry, I seemed to made the wrong statement about the behavior earlier -- DAG processor updates the metadata database, and scheduler tries to infer updated state from database as well, they're not making information exchange about the DAGs directly. When running standalone parser, scheduler look up the timestamp and stale duration to decide if some DAGs are out-of-sync, and deactivate them as needed.

So perhaps there are some job overlapping between processor and scheduler, but I'm not certain about the assumption/idea behind the design (for instance, should all the DAGs eventually stop and become deactivated when DAG processor's down? If it should, then letting scheduler take over DAG processor's responsibility may be reasonable).

**Reply by pykenny** [reply]
As for duplicated work on DAG processors, there were some discussion about how to avoid such situation, and I feel like it's still an open question for now:
 - https://github.com/apache/airflow/issues/32966
 
(Update) I think distributing workload between multiple DAG processors evenly without having an outside organizer (for instance, scheduler) would end up with a distributed system design problem (e.g. introducing consensus algorithm to decide when to re-distribute job load when one of the processors is down). Otherwise something like randomized sorting or jittering may help with preventing, but not avoiding the worst scenario to happen.
